langchain/schema/runnable.py

```diff
@@ -0,0 +1,123 @@
+from abc import ABC, abstractmethod
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+from typing import (
+    Any,
+    AsyncIterator,
+    Coroutine,
+    Dict,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    TypedDict,
+    TypeVar,
+    Union,
+)
+
+from typing_extensions import Unpack
+
+from langchain.callbacks.manager import Callbacks
+
+from langchain.load.serializable import Serializable
+
+
+async def _gated_coro(semaphore: asyncio.Semaphore, coro: Coroutine) -> Any:
+    async with semaphore:
+        return await coro
+
+
+async def _gather_with_concurrency(n: Union[int, None], *coros: Coroutine) -> list:
+    if n is None:
+        return await asyncio.gather(*coros)
+
+    semaphore = asyncio.Semaphore(n)
+
+    return await asyncio.gather(*(_gated_coro(semaphore, c) for c in coros))
+
+
+class RunnableConfig(TypedDict, total=False):
+    """
+    Tags for this call and any sub-calls (eg. a Chain calling an LLM).
+    You can use these to filter calls.
+    """
+
+    tags: List[str]
+
+    """
+    Metadata for this call and any sub-calls (eg. a Chain calling an LLM).
+    Keys should be strings, values should be JSON-serializable.
+    """
+    metadata: Dict[str, Any]
+
+    """
+    Callbacks for this call and any sub-calls (eg. a Chain calling an LLM).
+    Tags are passed to all callbacks, metadata is passed to handle*Start callbacks.
+    """
+    callbacks: Callbacks
+
+
+Input = TypeVar("Input")
+# Output type should implement __concat__, as eg str, list, dict do
+Output = TypeVar("Output")
+
+
+class Runnable(Generic[Input, Output], ABC):
+    @abstractmethod
+    def invoke(self, input: Input, **kwargs: Unpack[RunnableConfig]) -> Output:
```

Comment:
What if we called these `.next()` and `.anext()`? Dodges the similar sounding naming of `.invoke` vs. `call` vs. `generate` and makes it clearer that it's just meant to be used as an iterator.